src/common/cockpitwebserver.c

```diff
@@ -1296,116 +1222,96 @@ on_incoming (GSocketService *service,
 
 /* ---------------------------------------------------------------------------------------------------- */
 
-static gboolean
-cockpit_web_server_initable_init (GInitable *initable,
-                                  GCancellable *cancellable,
-                                  GError **error)
+gboolean
+cockpit_web_server_add_socket (CockpitWebServer *self,
+                               GSocket *socket,
+                               GError **error)
 {
-  CockpitWebServer *server = COCKPIT_WEB_SERVER (initable);
-  GSocketAddress *socket_address = NULL;
-  GSocketAddress *result_address = NULL;
-
-  gboolean ret = FALSE;
-  gboolean failed = FALSE;
-  int n, fd;
-
-  server->socket_service = g_socket_service_new ();
-
-  /* The web server has to be explicitly started */
-  g_socket_service_stop (server->socket_service);
+  return g_socket_listener_add_socket (G_SOCKET_LISTENER (self->socket_service), socket, NULL, error);
+}
 
-  n = sd_listen_fds (0);
-  if (n > 0)
+guint16
+cockpit_web_server_add_inet_listener (CockpitWebServer *self,
+                                      const gchar *address,
+                                      guint16 port,
+                                      GError **error)
+{
+  if (address != NULL)
     {
-      /* We got file descriptors passed in, use those. */
-
-      for (fd = SD_LISTEN_FDS_START; fd < SD_LISTEN_FDS_START + n; fd++)
+      g_autoptr(GSocketAddress) socket_address = g_inet_socket_address_new_from_string (address, port);
```

Comment:
It's legit to call that with `0`? https://docs.gtk.org/gio/ctor.InetSocketAddress.new_from_string.html does not mention that case.

Reply:
It must be.  `0` is a completely legit port number to have in an inet socket address structure, with a well-defined meaning: assign a port for me.  Also: it works :)

Note that the old code used to call

```
socket_address = g_inet_socket_address_new (server->address, server->port);
```

with a potentially-zero port.

Reply:
Tought as much, but good to confirm. Thanks!